contracts/strategies/direct-grants-simple/DirectGrantsSimpleStrategy.sol

```diff
@@ -0,0 +1,388 @@
+    function withdraw(uint256 _amount) external onlyPoolManager(msg.sender) {
+        allo.decreasePoolTotalFunding(poolId, _amount);
+        _transferAmount(allo.getPool(poolId).token, msg.sender, _amount);
+    }
+
+    /// ====================================
+    /// ============ Internal ==============
+    /// ====================================
+
+    /// @notice Register to the pool
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the transaction
+    function _registerRecipient(bytes memory _data, address _sender)
+        internal
+        override
+        onlyActivePool
+        returns (address recipientId)
+    {
+        address recipientAddress;
+        bool isRegistryIdentity;
+        uint256 grantAmount;
+        Metadata memory metadata;
+
+        // decode data custom to this strategy
+        if (registryGating) {
+            (recipientId, recipientAddress, grantAmount, metadata) =
+                abi.decode(_data, (address, address, uint256, Metadata));
+
+            if (!_isIdentityMember(recipientId, _sender)) {
+                revert UNAUTHORIZED();
+            }
+        } else {
+            (recipientAddress, isRegistryIdentity, grantAmount, metadata) =
+                abi.decode(_data, (address, bool, uint256, Metadata));
+            recipientId = _sender;
+            if (isRegistryIdentity && !_isIdentityMember(recipientId, _sender)) {
+                revert UNAUTHORIZED();
+            }
+        }
+
+        if (grantAmountRequired && grantAmount == 0) {
+            revert INVALID_REGISTRATION();
+        }
+        if (upcomingMilestone[recipientId] != 0) {
+            revert RECIPIENT_ALREADY_ACCEPTED();
+        }
+
+        Recipient memory recipient = Recipient({
+            recipientAddress: recipientAddress,
+            isRegistryIdentity: registryGating ? true : isRegistryIdentity,
+            grantAmount: grantAmount,
+            metadata: metadata,
+            recipientStatus: InternalRecipientStatus.Pending
+        });
+
+        _recipients[recipientId] = recipient;
+
+        emit Registered(recipientId, _data, _sender);
+    }
+
+    /// @notice Allocate amount to recipent for direct grants
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the allocation
+    function _allocate(bytes memory _data, address _sender)
+        internal
+        virtual
+        override
+        nonReentrant
+        onlyPoolManager(_sender)
+    {
+        (address recipientId, RecipientStatus recipientStatus, uint256 grantAmount) =
+            abi.decode(_data, (address, RecipientStatus, uint256));
+
+        if (upcomingMilestone[recipientId] != 0) {
+            revert MILESTONES_ALREADY_SET();
+        }
+
+        Recipient storage recipient = _recipients[recipientId];
+
+        if (recipientStatus == RecipientStatus.Accepted) {
+            recipient.grantAmount = grantAmount;
```

Comment:
should we track the `totalGrantAmount` globally to ensure that the `totalGrantAmount <= pool.amount` ?

Reply:
makes sense. calling it _allocatedGrantAmount

Reply:
can you add a getter for `_allocatedGrantAmount` so the pool manager can always check how much is left

Reply:
made it into a public variable